homeassistant/components/rachio.py

```diff
@@ -0,0 +1,282 @@
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['rachiopy==0.1.3']
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'rachio'
+
+CONF_CUSTOM_URL = 'hass_url_override'
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_API_KEY): cv.string,
+        vol.Optional(CONF_CUSTOM_URL): cv.string,
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+# Keys used in the API JSON
+KEY_DEVICE_ID = 'deviceId'
+KEY_DEVICES = 'devices'
+KEY_ENABLED = 'enabled'
+KEY_EXTERNAL_ID = 'externalId'
+KEY_ID = 'id'
+KEY_NAME = 'name'
+KEY_ON = 'on'
+KEY_STATUS = 'status'
+KEY_SUBTYPE = 'subType'
+KEY_SUMMARY = 'summary'
+KEY_TYPE = 'type'
+KEY_URL = 'url'
+KEY_USERNAME = 'username'
+KEY_ZONE_ID = 'zoneId'
+KEY_ZONE_NUMBER = 'zoneNumber'
+KEY_ZONES = 'zones'
+
+STATUS_ONLINE = 'ONLINE'
+STATUS_OFFLINE = 'OFFLINE'
+
+# Device webhook values
+TYPE_CONTROLLER_STATUS = 'DEVICE_STATUS'
+SUBTYPE_OFFLINE = 'OFFLINE'
+SUBTYPE_ONLINE = 'ONLINE'
+SUBTYPE_OFFLINE_NOTIFICATION = 'OFFLINE_NOTIFICATION'
+SUBTYPE_COLD_REBOOT = 'COLD_REBOOT'
+SUBTYPE_SLEEP_MODE_ON = 'SLEEP_MODE_ON'
+SUBTYPE_SLEEP_MODE_OFF = 'SLEEP_MODE_OFF'
+SUBTYPE_BROWNOUT_VALVE = 'BROWNOUT_VALVE'
+SUBTYPE_RAIN_SENSOR_DETECTION_ON = 'RAIN_SENSOR_DETECTION_ON'
+SUBTYPE_RAIN_SENSOR_DETECTION_OFF = 'RAIN_SENSOR_DETECTION_OFF'
+SUBTYPE_RAIN_DELAY_ON = 'RAIN_DELAY_ON'
+SUBTYPE_RAIN_DELAY_OFF = 'RAIN_DELAY_OFF'
+
+# Schedule webhook values
+TYPE_SCHEDULE_STATUS = 'SCHEDULE_STATUS'
+SUBTYPE_SCHEDULE_STARTED = 'SCHEDULE_STARTED'
+SUBTYPE_SCHEDULE_STOPPED = 'SCHEDULE_STOPPED'
+SUBTYPE_SCHEDULE_COMPLETED = 'SCHEDULE_COMPLETED'
+SUBTYPE_WEATHER_NO_SKIP = 'WEATHER_INTELLIGENCE_NO_SKIP'
+SUBTYPE_WEATHER_SKIP = 'WEATHER_INTELLIGENCE_SKIP'
+SUBTYPE_WEATHER_CLIMATE_SKIP = 'WEATHER_INTELLIGENCE_CLIMATE_SKIP'
+SUBTYPE_WEATHER_FREEZE = 'WEATHER_INTELLIGENCE_FREEZE'
+
+# Zone webhook values
+TYPE_ZONE_STATUS = 'ZONE_STATUS'
+SUBTYPE_ZONE_STARTED = 'ZONE_STARTED'
+SUBTYPE_ZONE_STOPPED = 'ZONE_STOPPED'
+SUBTYPE_ZONE_COMPLETED = 'ZONE_COMPLETED'
+SUBTYPE_ZONE_CYCLING = 'ZONE_CYCLING'
+SUBTYPE_ZONE_CYCLING_COMPLETED = 'ZONE_CYCLING_COMPLETED'
+
+# Webhook callbacks
+LISTEN_EVENT_TYPES = ['DEVICE_STATUS_EVENT', 'ZONE_STATUS_EVENT']
+WEBHOOK_CONST_ID = 'homeassistant.rachio'
+WEBHOOK_PATH = URL_API + DOMAIN
+EVENT_RACHIO_UPDATE = DOMAIN + '_update'
+EVENT_RACHIO_CONTROLLER_UPDATE = EVENT_RACHIO_UPDATE + '_controller'
+EVENT_RACHIO_ZONE_UPDATE = EVENT_RACHIO_UPDATE + '_zone'
+EVENT_RACHIO_SCHEDULE_UPDATE = EVENT_RACHIO_UPDATE + '_schedule'
+
+
+def setup(hass, config) -> bool:
+    """Set up the Rachio component."""
+    from rachiopy import Rachio
+
+    # Listen for incoming webhook connections
+    hass.http.register_view(RachioWebhookView())
+
+    # Configure API
+    api_key = config[DOMAIN].get(CONF_API_KEY)
+    rachio = Rachio(api_key)
+
+    # Get the URL of this server
+    custom_url = config[DOMAIN].get(CONF_CUSTOM_URL)
```

Comment:
When would a custom url be necessary ?

Reply:
For cases where the `base_url` will not be accessible by Rachio's server, or is incorrect. I don't know how commonly this will happen, but I figured it should be possible after there were [similar issues with another component](https://community.home-assistant.io/t/doorbird-integration/5564/67).